src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -100,13 +100,13 @@ private PerAnalyzerState GetAnalyzerState(DiagnosticAnalyzer analyzer)
         {
             await EnsureAnalyzerActionCountsInitializedAsync(driver, cancellationToken).ConfigureAwait(false);
 
-            lock (_gate)
+            using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
             {
-                OnCompilationEventsGenerated_NoLock(compilationEvents, driver, cancellationToken);
+                await OnCompilationEventsGenerated_NoLockAsync(compilationEvents, driver, cancellationToken).ConfigureAwait(false);
             }
         }
 
-        private void OnCompilationEventsGenerated_NoLock(ImmutableArray<CompilationEvent> compilationEvents, AnalyzerDriver driver, CancellationToken cancellationToken)
+        private async Task OnCompilationEventsGenerated_NoLockAsync(ImmutableArray<CompilationEvent> compilationEvents, AnalyzerDriver driver, CancellationToken cancellationToken)
         {
             Debug.Assert(_lazyAnalyzerActionCountsMap != null);
 
```

Comment:
_lazyAnalyzerActionCountsMap is not readonly so checking whether it is not null but yield the execution for other code to run can change the field. that is why I said null checking there might no longer valid. but, code itself could guarantee that doesnt happen. so in that case it is fine.

same goes for _pooledEvents ..., it is readonly but correction is mutable. so while task is pending (when other code runs), the entry in the collection could change.